core: URL Callbacks new interface

Exirel opened the pull request:
This new interface adds:

* `bot.register_url_callback(regex, callback)`, to register a `callback` to call when a URL matches `regex`
* `bot.unregister_url_callback(regex)` to remove this regex from triggering callbacks
* `bot.search_url_callbacks(url)` to get a list of (callback, match) for each pattern matched by the URL

This provides a unified interface to work with URL Callbacks, but does not change how the built-in url plugin works (yet).

So this:

    if not bot.memory.contains('url_callbacks'):
        bot.memory['url_callbacks'] = tools.SopelMemory()
    bot.memory['url_callbacks'][instagram_pattern] = instaparse

Become this:

    bot.register_url_callback(instagram_pattern, instaparse)

It does not change the fact that the `url` built-in plugin is required at the moment in order to work with URL Callback (see #1489 for that).

@dgw now my question is, should I keep moving forward, and replace the `url` built-in plugin? Or should we merge this PR as-is, and work from there in another PR?

dgw commented:
You can forge on ahead with a new PR branch based on this one, like you've done before, unless you need to discuss direction (in #1489) first. That way, no matter how much bikeshedding happens with replacing the url module, we'll have a clean new interface in the bag.

I like that this really is just adding convenience methods to do what plugins already do, so nothing will break (yet) if module code isn't updated. 😁

Do you think it's worth checking in the register/unregister functions added here whether the `pattern` is actually compiled? I can picture users who merely skimmed the documentation (or coders drunk on lack of sleep XD) trying to pass in an uncompiled regex string…

dgw commented on sopel/bot.py:
```diff
@@ -648,3 +646,71 @@ def cap_req(self, module_name, capability, arg=None, failure_callback=None,
+        :param pattern: compiled regex pattern to register
+        :param callback: callable object to handle matching URLs
+
+        .. versionadded: 7.0.0
```
This directive won't appear in Sphinx's output as written; it needs two colons. (Also applies to the other two occurrences.)

I would also just put `7.0`, since we're not _ever_ (if I have anything to say about it) going to add/change/remove API methods in a patch release.

```suggestion
        .. versionadded:: 7.0
```

Exirel replied:
Fixed.

dgw commented on sopel/bot.py:
```diff
@@ -648,3 +646,71 @@ def cap_req(self, module_name, capability, arg=None, failure_callback=None,
+        .. versionadded: 7.0.0
+
+            This method has been added in Sopel 7.0.0 in order to offer
+            a unified interface for URL callbacks.
```
Redundant to say "This method has been added in Sopel 7.0.0"—the `versionadded` directive already says that for us. (Applies to all three added docstrings.)

This explanatory text is OK, but I think it would be more useful to show module authors what code it replaces in their `setup` functions (the old `bot.memory` check and subsequent manual assignment into the dict).

Exirel replied:
Done.

dgw commented on sopel/bot.py:
```diff
@@ -648,3 +646,71 @@ def cap_req(self, module_name, capability, arg=None, failure_callback=None,
+        .. versionadded: 7.0.0
+
+            This method has been added in Sopel 7.0.0 in order to offer
+            a unified interface to URL callbacks.
```
Similar to the `register_` function docstring, it would be useful here to show module authors what to replace in their `shutdown` methods.

Exirel replied:
Done.

dgw commented on sopel/bot.py:
```diff
@@ -648,3 +646,71 @@ def cap_req(self, module_name, capability, arg=None, failure_callback=None,
+        .. versionadded: 7.0.0
+
+            This method has been added in Sopel 7.0.0 in order to offer
+            a unified interface for URL callbacks.
```
Not sure what would be more useful to put here, but the comment I made about redundancy on `register_` above applies here. Maybe this one doesn't need an explanation for the `versionadded` directive at all?

Exirel replied:
Removed explanation - done.

Exirel commented:
> I like that this really is just adding convenience methods to do what plugins already do, so nothing will break (yet) if module code isn't updated. 😁

I knew you would understand. 👍 

> Do you think it's worth checking in the register/unregister functions added here whether the `pattern` is actually compiled?

Yeah, I was wondering too. It's not that complicated, I'll do that.

Exirel commented:
Also: of course I failed at docstring again. x)

Exirel commented:
Hm. I'm not sure "re-request a review" was the right idea. Meh, I guess I need to learn what the new github has to offer with all these features.

Anyway @dgw I fixed the docstrings and added the safety measures to check if pattern is a compiled regex.

Exirel commented:
Ahem. Checkstyle?

dgw commented on sopel/bot.py:
```diff
@@ -648,3 +646,97 @@ def cap_req(self, module_name, capability, arg=None, failure_callback=None,
+                regex = re.compile(r'http://example.com/path/.*')
+                bot.memory['url_callbacks'][regex] = callback
+
+            Plugin's maintainers must do this::
```
"must" is a bit strong?

```suggestion
            use this much more concise pattern::
```

(starts with lowercase because it's really continuing the previous sentence)

Also applies to `unregister_`.

Exirel replied:
I like your version. I never know how to properly use "must". :(

Exirel commented on sopel/test_tools.py:
```diff
@@ -74,6 +75,12 @@ def _init_config(self):
             os.mkdir(home_dir)
         cfg.parser.set('core', 'homedir', home_dir)
 
+    def search_url_callbacks(self, url):
```
I honestly don't know why. :-/

Exirel replied:
Ok, I totally need to add the register/unregister method too, because it might fail in the future.

It's complicated, but TL;DR: it works now because none of the plugin that use register/unregister have test example, so their setup function is never called, so these methods are never called. However, the URL module is tested, and it uses the search method, so it needs to be here.

So, right now, it's pure luck that it works. I'll fix that and make sure it works *because that's the right thing to do*.

dgw replied:
OK, now we're adding three of `bot`'s methods to the `MockSopel` but not any of the others that modules might call. Something feels very wrong about this.

`say`, `reply`, and `action` are defined in `MockSopelWrapper` to store the command's output:

https://github.com/sopel-irc/sopel/blob/d2e7a163ff06e7697e857e77b14681cac251c845/sopel/test_tools.py#L87

We're not already looking at test problems because, for example, none of the core modules that use `bot.db` have testable examples. If they did have example results, errors would happen.

Same for `notice`—no core module command both has a testable example and calls `bot.notice()` while doing its work. If that happened, error.

I'm not going to "Unresolve" this conversation, because doing anything about any of this would be way out of scope for this PR, but… Basically, next time you feel ambitious about making the _tests_ better, it would be worth finding an alternative solution, like using the real Sopel but mocking out only its connection to IRC instead. Keeping the "real" and "mock" implementations/APIs in sync is only going to get sillier as we add features to the API and try to improve test coverage of modules.

Exirel replied:
I fully agree with you. I'm not happy with these mocks, and they don't cover everything we need.

As said on IRC, we need time to figure something out that will be better and fit our needs. I'm sure we just need some time for this to sink in and come back to that topic with fresh ideas.

dgw commented:
Merging so @Exirel can properly get started on a PR to fix the mess that is `url.py`. :vulcan_salute:
